### ISSUE TITLE
[BUG] Conditions SC2034, SC2102 trigger on valid code

### ISSUE THREAD

**Issue by r6d2** [koalaman/shellcheck#1308]
#### For bugs
- Rule Id: SC2034, SC2102
- My shellcheck version: online
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and found it may be related to issue #1107.

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/usr/bin/env bash
        texture=ABC
	[[ -v PATH_OF_TEXTURE[$texture] ]] && color=lightblue	# color for MAP_TEXTURE_PATCH
	echo "$color"

```

#### Here's what shellcheck currently says:
```
Line 2	SC2034: texture appears unused. Verify use (or export if used externally).
Line 3	SC2102: Ranges can only match single chars (mentioned due to duplicates).
```
#### Here's what I wanted or expected to see:

Nothing. Apparently, shellcheck is unable to grasp the usage of variable $texture as an index of an array (says the variable is unused), and also seems to believe [] is always a range and not a reference on an array.

**Comment by matthewpersico** [comment]
What are you expecting [[ -v PATH_OF_TEXTURE[$texture] ]] to do? If PATH_OF_TEXTURE is supposed to be an associative array, don't you need to declare -A PATH_OF_TEXTURE and then refer to it as ${PATH_OF_TEXURE[$texture]}?

**Comment by r6d2** [comment]
_[I'm sorry I'm getting back to your question on this so late, please accept my apologies.]_

I expect that shellcheck would consider `[[ -v PATH_OF_TEXTURE[$texture] ]]` as a valid form of verifying the following (from the bash reference manual):

> -v varname
> True if the shell variable varname is set (has been assigned a value).

In this case `PATH_OF_TEXTURE[$texture]` is the varname. AFAIK your suggestion to use `${PATH_OF_TEXURE[$texture]}` instead should consider the value stored in the variable for testing purposes and nor the variable name.

Anyway, what I did as a workaround was to write the code like this:
`[[ -v "PATH_OF_TEXTURE[$texture]" ]]`

I get the desired result and no warnings from shellcheck.

**Comment by cfstras** [comment]
Having the same problem. Here's a canonical reproducer:

```bash
#!/bin/bash

declare -A associative_array=(
  ["key"]="123"
)

get_value() {
  local key_variable="${1}"
  if [[ ! -v associative_array[$key_variable] ]]; then
    echo "not present"
  else
    echo "${associative_array[$key_variable]}"
  fi
}
```

```
  if [[ ! -v associative_array[$key_variable] ]]; then
                              ^-- SC2102: Ranges can only match single chars (mentioned due to duplicates).
```

Interestingly, it depends on the actual name of `key_variable`. Using `key` and `key_var` does not error.
